pkg/prometheus/promcfg.go

```diff
@@ -554,22 +565,25 @@ func (cg *ConfigGenerator) Generate(
 		Value: append(scrapeConfigs, addlScrapeConfigs...),
 	})
 
-	cfg, err = cg.appendStorageSettingsConfig(cfg, p)
-	if err != nil {
-		return nil, errors.Wrap(err, "generating storage_settings configuration failed")
+	if len(cpf.RemoteWrite) > 0 {
+		cfg = append(cfg, cg.generateRemoteWriteConfig(p.GetCommonPrometheusFields(), store, p.GetObjectMeta().Namespace))
 	}
 
-	cfg, err = cg.appendAlertingConfig(cfg, p, additionalAlertRelabelConfigs, additionalAlertManagerConfigs, store)
-	if err != nil {
-		return nil, errors.Wrap(err, "generating alerting configuration failed")
-	}
+	switch prom := p.(type) {
```

Comment:
I'm not a fan of type casting in Go :)
Since alerting isn't applicable to the Prometheus agent, I'd suggest to split the Generate() function into more composable fragments. The server and agent controllers can pipe them up as needed. 

```golang
func (cg *ConfigGenerator) AppendScrapeConfigs(
  cfg yaml.MapSlice,
  ..., // required inputs for building the scrape_config
) (yaml.MapSlice, error) {
  ...
}

func (cg *ConfigGenerator) AppendAlertingConfig(
  cfg yaml.MapSlice,
...
) (yaml.MapSlice, error) {
  ...
}

func (cg *ConfigGenerator) AppendGlobalConfig(
  cfg yaml.MapSlice,
...
) (yaml.MapSlice, error) {
  ...
}

func (cg *ConfigGenerator) AppendRemoteWrite(
  cfg yaml.MapSlice,
...
) (yaml.MapSlice, error) {
  ...
}
```

Reply:
I don't see how we avoid typecasting with this approach 🤔

If both controllers use the same ConfigGenerator, we need to know the resource type to decide if we're calling server-specific configuration or not, don't we?

Reply:
For the record, my idea is that each controller will build its own Generate() function based on primitives offered by the ConfigGenerator struct.